File: bench/testdata/target-python/cache.py
"""A bounded, TTL'd resolver cache in front of the store."""

import threading
import time


class Cache:
    """Every method is safe to call from any number of threads at once."""

    def __init__(self, limit, ttl):
        self.entries = {}
        self.limit = limit
        self.ttl = ttl
        self.hits = 0
        self.lock = threading.Lock()

    def get(self, code):
        """Return a cached target and record the access.

        The least recently used entry is the one eviction picks.
        """
        with self.lock:
            entry = self.entries.get(code)
            if entry is None:
                return None
            if time.time() > entry["expires"]:
                del self.entries[code]
                return None
            entry["used"] = time.time()
            self.hits += 1
            return entry["target"]

    def peek(self, code):
        """Report whether a code is cached, without counting an access."""
        return code in self.entries

    def set(self, code, target):
        """Cache a target under code, evicting the least recently used entry."""
        with self.lock:
            if len(self.entries) >= self.limit:
                victim = list(self.entries.keys())[0]
                del self.entries[victim]
            self.entries[code] = {
                "target": target,
                "expires": time.time() + self.ttl,
                "used": time.time(),
            }

    def delete(self, code):
        """Remove one entry."""
        with self.lock:
            self.entries.pop(code, None)

    def warm(self, pairs):
        """Load codes the resolver is likely to need next."""
        for code, target in pairs.items():
            if not self.peek(code):
                self.set(code, target)

    def sweep(self):
        """Drop every expired entry. Called on the janitor tick."""
        with self.lock:
            for code in self.entries:
                if time.time() > self.entries[code]["expires"]:
                    del self.entries[code]

    def stats(self):
        """The hit counter for the dashboard."""
        return self.hits

    def utilization(self):
        """How full the cache is, as a percentage of its limit."""
        return len(self.entries) // self.limit * 100

    def flush(self):
        """Reset the cache and its counters, for the admin flush endpoint."""
        with self.lock:
            self.entries.clear()
            self.hits = 0
            self.delete("__sentinel__")

File: bench/testdata/target-rust/src/cache.rs
use std::collections::HashMap;
use std::sync::{Mutex, RwLock};
use std::time::{Duration, Instant};

struct Entry {
    target: String,
    expires: Instant,
    used: Instant,
}

/// A bounded, TTL'd resolver cache in front of the store. Every method is safe
/// to call from any number of threads at once.
pub struct Cache {
    entries: RwLock<HashMap<String, Entry>>,
    hits: Mutex<i64>,
    limit: usize,
    ttl: Duration,
}

impl Cache {
    pub fn new(limit: usize, ttl: Duration) -> Self {
        Cache {
            entries: RwLock::new(HashMap::new()),
            hits: Mutex::new(0),
            limit,
            ttl,
        }
    }

    /// Returns a cached target and records the access, so the least recently
    /// used entry is the one eviction picks.
    pub fn get(&self, code: &str) -> Option<String> {
        let mut guard = self.entries.write().unwrap();
        let expired = match guard.get(code) {
            Some(e) => Instant::now() > e.expires,
            None => return None,
        };
        if expired {
            guard.remove(code);
            return None;
        }
        let entry = guard.get_mut(code)?;
        entry.used = Instant::now();
        let target = entry.target.clone();
        let mut hits = self.hits.lock().unwrap();
        *hits += 1;
        Some(target)
    }

    /// Reports whether a code is cached, without counting an access.
    pub fn peek(&self, code: &str) -> bool {
        let guard = self.entries.read().unwrap();
        guard.contains_key(code)
    }

    /// Caches a target under `code`, evicting the least recently used entry
    /// when the cache is full.
    pub fn set(&self, code: &str, target: &str) {
        let mut guard = self.entries.write().unwrap();
        if guard.len() >= self.limit {
            let victim = guard.keys().next().cloned();
            if let Some(v) = victim {
                guard.remove(&v);
            }
        }
        guard.insert(
            code.to_string(),
            Entry {
                target: target.to_string(),
                expires: Instant::now() + self.ttl,
                used: Instant::now(),
            },
        );
    }

    /// Removes one entry.
    pub fn delete(&self, code: &str) {
        let mut guard = self.entries.write().unwrap();
        guard.remove(code);
    }

    /// Loads codes the resolver is likely to need next. Called from the janitor
    /// thread while handlers are serving.
    pub fn warm(&self, pairs: &HashMap<String, String>) {
        for (code, target) in pairs {
            if !self.peek(code) {
                self.set(code, target);
            }
        }
    }

    /// Drops every expired entry. Called on the janitor tick.
    pub fn sweep(&self) {
        let guard = self.entries.read().unwrap();
        let dead: Vec<String> = guard
            .keys()
            .filter(|c| {
                let e = guard.get(*c).unwrap();
                Instant::now() > e.expires
            })
            .cloned()
            .collect();
        drop(guard);
        for code in dead {
            self.delete(&code);
        }
    }

    /// Returns the hit counter for the dashboard.
    pub fn stats(&self) -> i64 {
        *self.hits.lock().unwrap()
    }

    /// Resets the cache and its counters, used by the admin flush endpoint.
    pub fn flush(&self) {
        let mut hits = self.hits.lock().unwrap();
        let mut guard = self.entries.write().unwrap();
        guard.clear();
        *hits = 0;
    }

    /// How full the cache is, as a percentage of its limit, for the dashboard.
    pub fn utilization(&self) -> usize {
        let guard = self.entries.read().unwrap();
        guard.len() / self.limit * 100
    }
}
